zirgen/Dialect/ZHLT/Transforms/ElideRedundantMembers.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "zirgen/Dialect/ZHLT/IR/ZHLT.h"
+#include "zirgen/Dialect/ZHLT/Transforms/PassDetail.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/ADT/Statistic.h"
+
+#define DEBUG_TYPE "elide-redundant-members"
+
+using namespace mlir;
+using namespace zirgen::ZStruct;
+
+namespace zirgen::Zhlt {
+
+template <typename StructLikeType>
+StructLikeType dropMembers(StructLikeType type, ArrayRef<std::pair<size_t, size_t>> merges) {
+  SmallVector<FieldInfo> fields(type.getFields());
+  for (auto merge : llvm::reverse(merges)) {
+    fields.erase(fields.begin() + merge.second);
+  }
+  return StructLikeType::get(type.getContext(), type.getId(), fields);
+}
+
+// For structure-like components, if two members are equal in the PackOp at the
+// end of the constructor, those members will ultimately be equal in all other
+// situations, such as when reconstructing an instance from a back and when
+// zero-initializing (trivially, since both members are zeroed).
+struct ElideRedundantMembersPass : public ElideRedundantMembersBase<ElideRedundantMembersPass> {
+  ElideRedundantMembersPass() = default;
+  ElideRedundantMembersPass(const ElideRedundantMembersPass& pass) {}
+
+  struct Data {
+    Type newType;
+    SmallVector<std::pair<size_t, size_t>> merges;
+  };
+
+  void runOnOperation() override {
+    DenseMap<Type, Data> mapping;
+    for (auto component : getOperation().getBodyRegion().getOps<ComponentOp>()) {
+      auto ret = cast<ReturnOp>(component.getBody().back().getTerminator());
+      auto pack = dyn_cast_if_present<PackOp>(ret.getValue().getDefiningOp());
+      if (!pack)
+        continue;
+
+      StructType type = pack.getOut().getType();
+      auto members = pack.getMembers();
+      SmallVector<std::pair<size_t, size_t>> merges;
+      for (size_t i = 0; i < members.size(); i++) {
+        for (size_t j = i + 1; j < members.size(); j++) {
+          if (members[i] == members[j]) {
```

Comment:
The quadratic here gives me pause, though perhaps we can expect N to be small enough that it won't matter. I do wonder if it could be done more efficiently with a multiset, or more likely unordered_multiset: loop once over the member list to map the location of each member, then loop again over the multiset to spot duplicates.